apps/els_core/src/els_provider.erl

```diff
@@ -43,78 +43,134 @@
+                             }.
+-type job() :: pid().
+%% TODO: Redefining uri() due to a type conflict with request()
+-type uri() :: binary().
 -export_type([ config/0
              , provider/0
              , request/0
              , state/0
              ]).
 
+%%==============================================================================
+%% Macro Definitions
+%%==============================================================================
+-define(SERVER, ?MODULE).
+
 %%==============================================================================
 %% External functions
 %%==============================================================================
 
--spec start_link(provider()) -> {ok, pid()}.
-start_link(Provider) ->
-  gen_server:start_link({local, Provider}, ?MODULE, Provider, []).
+-spec start_link() -> {ok, pid()}.
+start_link() ->
+  gen_server:start_link({local, ?SERVER}, ?MODULE, unused, []).
 
 -spec handle_request(provider(), request()) -> any().
 handle_request(Provider, Request) ->
-  gen_server:call(Provider, {handle_request, Request}, infinity).
+  gen_server:call(?SERVER, {handle_request, Provider, Request}, infinity).
+
+-spec cancel_request(pid()) -> any().
+cancel_request(Job) ->
+  gen_server:cast(?SERVER, {cancel_request, Job}).
 
--spec cancel_request(provider(), pid()) -> any().
-cancel_request(Provider, Job) ->
-  gen_server:cast(Provider, {cancel_request, Job}).
+-spec cancel_request_by_uri(uri()) -> any().
+cancel_request_by_uri(Uri) ->
+  gen_server:cast(?SERVER, {cancel_request_by_uri, Uri}).
 
 %%==============================================================================
 %% gen_server callbacks
 %%==============================================================================
 
--spec init(els_provider:provider()) -> {ok, state()}.
-init(Provider) ->
-  ?LOG_INFO("Starting provider ~p", [Provider]),
-  InternalState = case erlang:function_exported(Provider, init, 0) of
-                    true ->
-                      Provider:init();
-                    false ->
-                      #{}
-                  end,
-  {ok, #{provider => Provider, internal_state => InternalState}}.
+-spec init(unused) -> {ok, state()}.
+init(unused) ->
+  {ok, #{in_progress => [], in_progress_diagnostics => []}}.
 
 -spec handle_call(any(), {pid(), any()}, state()) ->
   {reply, any(), state()}.
-handle_call({handle_request, Request}, _From, State) ->
-  #{internal_state := InternalState, provider := Provider} = State,
-  {Reply, NewInternalState} = Provider:handle_request(Request, InternalState),
-  {reply, Reply, State#{internal_state => NewInternalState}}.
+handle_call({handle_request, Provider, Request}, _From, State) ->
+  #{in_progress := InProgress, in_progress_diagnostics := InProgressDiagnostics}
+    = State,
+  case Provider:handle_request(Request, State) of
+    {async, Uri, Job} ->
+      {reply, {async, Job}, State#{in_progress => [{Uri, Job}|InProgress]}};
```

Comment:
Is there some monitoring of the jobs to remove them from the map when they are externally terminated/crash?

Reply:
Not that I can see, even if the consequence of leaving an orphan entry in the map shouldn't be critical. I will create a ticket about this. Also, I now realize that there's some logic running in the `terminate` function (to clean up the spinning wheel), so I will roll back the `brutal_kill` strategy bit.

Reply:
#1271